### ISSUE TITLE
[BUG] Indirect lineage requires an output field

### ISSUE THREAD

**Issue by Sheeri** [OpenLineage/OpenLineage#2799]
### OpenLineage integration/client

Other

### OpenLineage version

spec 1-1-0, specifically the changes in #2756

### Technology and package versions

N/A

### Environment configuration

N/A

### Deployment details

N/A

### Problem details

I tried to update the test case for https://github.com/OpenLineage/OpenLineage/blob/main/spec/tests/ColumnLineageDatasetFacet/1.json based on the changes in #2756 (and its relation in closing #2186)

I have a few questions:
1) How should we specify spec version in the test case? I can make 2.json but how to specify that 1.json is for spec version 1-0-2 and 2.json is for 1-1-0?

2) The spec at https://github.com/OpenLineage/OpenLineage/blob/main/spec/facets/ColumnLineageDatasetFacet.json requires inputFields to have an [output]field as a parent.

However, INDIRECT lineage doesn't have a corresponding outputField by definition.

As an example, lineage from:
CREATE VIEW CUSTOMER_DISCOUNTS AS
SELECT NAME, AMOUNT_OFF, STARTS_AT, ENDS_AT
FROM DISCOUNTS INNER JOIN CUSTOMERS ON (DISCOUNTS.CUSTOMERS_ID=CUSTOMERS.ID)

The inputField CUSTOMERS_ID (from the DISCOUNTS table) has a transformationType of "INDIRECT", subType of "FILTER" but there's no parent field to put it under....

### What you think should happen instead

1. There should be a clear method to apply versions to test cases. e.g. we should be able to have 2 test cases, 1 used for spec version 1-0-2 that would have the transformation type as part of the outputDataset; and 1 used for spec version 1-1-0 that would have the transformation type (and subType) as part of the inputDataset.

This could be done with a version field in the test case, or it could be done by folders in the github repo, e.g.
spec/tests/1-1-0/ColumnLineageDatasetFacet/1.json
or
spec/tests/ColumnLineageDatasetFacet/1-1-0/1.json

or it could be done with a naming convention, e.g.

spec/tests/ColumnLineageDatasetFacet/1-0-2/1-1-0.1.json (1st test for spec version 1-0-2)
spec/tests/ColumnLineageDatasetFacet/1-0-2/1-1-0.2.json (2nd test for spec version 1-0-2)
spec/tests/ColumnLineageDatasetFacet/1-1-0/1-1-0.1.json (1st test for spec version 1-1-0)
spec/tests/ColumnLineageDatasetFacet/1-1-0/1-1-0.2.json (2nd test for spec version 1-1-0)

2. We should discuss how to apply indirect lineage to an inputDataset field and an outputDataset table object. There isn't an easy/obvious way to do this, though it could be done by putting the transformations array at the top level, e.g.:

`
  "columnLineage": {
    "transformations": [
              {
              "type": "INDIRECT",
              "subtype": "FILTER",
              "description": "ON (DISCOUNTS.CUSTOMERS_ID=CUSTOMERS.ID)",
              }
        ]
`

### How to reproduce

Try to make a test case with the 1-1-0 spec, for the indirect transformation:

CREATE VIEW CUSTOMER_DISCOUNTS AS
SELECT NAME, AMOUNT_OFF, STARTS_AT, ENDS_AT
FROM DISCOUNTS INNER JOIN CUSTOMERS ON (DISCOUNTS.CUSTOMERS_ID=CUSTOMERS.ID)

### Do you plan to make this contribution yourself?

- [X] I am interested in doing this work

**Comment by Sheeri** [comment]
FWIW I am super excited to have indirect lineage represented in OpenLineage! :D

**Comment by tnazarew** [comment]
@Sheeri 

> How should we specify spec version in the test case? I can make 2.json but how to specify that 1.json is for spec version 1-0-2 and 2.json is for 1-1-0?

In spec tests we only check the version we currently have in spec schema files. We don't check against previous versions, so if the version defined in code is 1-1-0, we only check test files against 1-1-0. 

> The spec at https://github.com/OpenLineage/OpenLineage/blob/main/spec/facets/ColumnLineageDatasetFacet.json requires inputFields to have an [output]field as a parent. 
> However, INDIRECT lineage doesn't have a corresponding outputField by definition.

**TL;DR** it does, in most cases, if there is INDIRECT dependency, all the output fields have it.

In general we can separate indirect dependencies into two categories, let's call them field dependencies and dataset dependencies. Field dependency means that that only one column depend on them, they have `CONDITIONAL` subtype. Dataset dependency means it affects all the outputs in the dataset, these are all the other subtypes (`JOIN`, `FILTER`, `GROUP_BY` etc.), so all the output columns have separately defined dependencies on them. 

Examples:

1. Field dependency 
```
CASE WHEN A > 1 THEN B ELSE C AS RES
``` 
then the `RES` output column has direct dependency on `B` and `C` but also indirect **_field_** dependency on `A`

2. Dataset dependency 
```
CREATE VIEW CUSTOMER_DISCOUNTS AS
SELECT A, B, C 
FROM T1 JOIN T2 ON (T1.A=T2.A)
WHERE B 
```
All of the columns depend indirectly on T1.A, T2.A and B, so each output column will have dependencies
T1.A -> INDIRECT JOIN
T2.A -> INDIRECT JOIN
B -> INDIRECT FILTER

**Comment by Sheeri** [comment]
OK. This works in theory; in practice it will get cumbersome for tables/views with multiple indirect dependencies, and lots of columns. e.g. a 10-column table with 2 indirect dependencies would generate 20 transformation JSON objects. Consumers can skip duplicates, and I guess extra network bandwidth isn't an issue?

I will make an example and close this issue with a PR for it.

**Comment by dolfinus** [comment]
> I guess extra network bandwidth isn't an issue

For HTTP transport, gzip compression can be used, it reduces body size a lot
